bufferserver/src/main/java/com/datatorrent/bufferserver/internal/LogicalNode.java

```diff
@@ -220,17 +220,17 @@ public void catchUp()
       }
 
       if (iterator.hasNext()) {
-        addedData();
+        addedData(false);
       }
     }
 
     logger.debug("Exiting catch up because caughtup = {}", caughtup);
   }
 
   @Override
-  public boolean addedData()
+  public boolean addedData(boolean checkIfListenerHaveDataToSendOnly)
   {
-    if (isReady()) {
+    if (!checkIfListenerHaveDataToSendOnly && isReady()) {
```

Comment:
Doesn't look like something that should be overloaded into this method. Why not make it a separate method to check if data is available to send and it accordingly to the interface.